astroid/helpers.py

```diff
@@ -314,3 +316,78 @@ def object_len(node, context=None):
     raise AstroidTypeError(
         f"'{result_of_len}' object cannot be interpreted as an integer"
     )
+
+
+def literal_eval(node_or_string: Union[str, nodes.NodeNG]) -> Any:
+    """
+    Safely evaluate an expression node or a string containing a Python
+    expression.  The string or node provided may only consist of the following
+    Python literal structures: strings, bytes, numbers, tuples, lists, dicts,
+    sets, booleans, and None.
+
+    Can raise `AstroidSyntaxError` or `ValueError`.
+
+    Adapted from `ast.literal_eval`.
+    """
+    if isinstance(node_or_string, str):
+        _node = builder.parse(node_or_string.lstrip(" \t")).body
+        if len(_node) != 1:
+            raise ValueError(f"expected only one expression, found {len(_node)}")
+        node_or_string = _node[0]
+    if isinstance(node_or_string, nodes.Expr):
+        node_or_string = node_or_string.value
+
+    def _raise_malformed_node(node):
+        msg = "malformed node or string"
+        lno = getattr(node, "lineno", None)
+        if lno:
+            msg += f" on line {lno}"
+        raise ValueError(msg + f": {node!r}")
+
+    def _convert_num(node):
+        if not isinstance(node, nodes.Const) or type(node.value) not in (
+            int,
+            float,
+            complex,
+        ):
+            _raise_malformed_node(node)
+        return node.value
+
+    def _convert_signed_num(node):
+        if isinstance(node, nodes.UnaryOp) and node.op in ("+", "-"):
+            operand = _convert_num(node.operand)
+            if node.op == "+":
+                return +operand
+            else:
+                return -operand
+        return _convert_num(node)
+
+    def _convert(node):
+        if isinstance(node, nodes.Const):
+            return node.value
+        elif isinstance(node, nodes.Tuple):
+            return tuple(map(_convert, node.elts))
+        elif isinstance(node, nodes.List):
+            return list(map(_convert, node.elts))
+        elif isinstance(node, nodes.Set):
+            return set(map(_convert, node.elts))
+        elif (
+            isinstance(node, nodes.Call)
+            and isinstance(node.func, nodes.Name)
+            and node.func.name == "set"
+            and node.args == node.keywords == []
+        ):
+            return set()
```

Comment:
`set()` is a special case because we can't create empty set other wise. 
I don;t think we should add this kind of support for `dict()` or `list()` since we can use `{}` and `[]`.

Reply:
Ah yeah. Still, if we're going to maintain our own function shouldn't we add some obvious extra functionality while we're at it?

Reply:
I get your point, we could handle empty calls to `list`, `dict`, `tuple`, `frozenset`, if it's only that, I'm ok implementing it. 

But in general, I think we should stay the closest possible from `ast` module, and let users call this function with a possible inferred node.

Reply:
Yeah, I think it makes sense to add those three as well. There is a big chance that we will run out of sync with the "upstream" version of this due to limited maintenance time. Makes sense imo to add some slight improvements while we're looking at it.